File: apps/web/src/lib/reality.ts
/**
 * Entity, conflict and timeline API client.
 *
 * Note what the types do and do not carry. `Conflict.score` is
 * `number | null` and `severity` can be `"unspecified"` — the confidence
 * specification is missing, so a conflict can be *detected* without being
 * *graded*. The interface must render that distinction rather than defaulting
 * a missing grade to something reassuring.
 */

import { apiFetch } from "@/lib/api";

export type ConflictType =
  "value_conflict" | "source_disagreement" | "contested_state";

/** "unspecified" means the severity thresholds are not available. */
export type ConflictSeverity =
  "low" | "medium" | "high" | "critical" | "unspecified";

export type ConflictStatus = "open" | "acknowledged" | "resolved" | "dismissed";

export type TimeAxis = "event" | "knowledge";

export interface Entity {
  id: string;
  entity_type: string;
  natural_key: string;
  display_name: string | null;
  mapping_count: number;
  observation_count: number;
  created_at: string;
}

export interface CompetingValue {
  value: unknown;
  weight: string;
  share: string;
  sources: string[];
  observation_count: number;
}

export interface Conflict {
  id: string;
  entity_id: string;
  entity_natural_key: string | null;
  reality_state_id: string | null;
  attribute: string;
  conflict_type: ConflictType;
  severity: ConflictSeverity;
  status: ConflictStatus;
  /** null while the conflict-score formula is unavailable. */
  score: string | null;
  summary: string;
  details: {
    competing_values?: CompetingValue[];
    divergence?: string | null;
    margin_percentage_points?: string;
    grading?: { available: boolean; reason?: string };
    [key: string]: unknown;
  };
  detected_at: string;
  last_seen_at: string;
  resolved_at: string | null;
  resolution_note: string | null;
}

export interface TimelineEvent {
  observation_id: string;
  external_id: string;
  source_id: string;
  source_name: string;
  values: Record<string, unknown>;
  /** When the fact was true, per the source. */
  event_time: string;
  /** When RealitySync learned it. */
  ingested_at: string;
  event_time_semantics: string;
  /** True when learned after it was true — the two axes diverged here. */
  arrived_late: boolean;
  lag_seconds: number;
}

export interface Timeline {
  axis: TimeAxis;
  as_of_event_time: string | null;
  as_of_knowledge_time: string | null;
  event_count: number;
  late_arrival_count: number;
  truncated: boolean;
  events: TimelineEvent[];
}

export interface RealityState {
  id: string;
  entity_id: string;
  attribute: string;
  /**
   * `null` when no value was selected. Two different situations, told apart by
   * `status`: `unknown` means no usable evidence exists, `contested` means
   * several values compete and the rule for ranking them is unavailable.
   */
  value: unknown;
  /**
   * Whether `value` is a selection. Read this rather than checking `value` for
   * null — a source can legitimately assert null, and the two must not be
   * rendered the same way.
   */
  value_selected: boolean;
  /** `null` while the confidence specification is unavailable. Never render 0. */
  confidence: string | null;
  /** Explicit flag, so the UI branches on intent rather than on a null. */
  confidence_available: boolean;
  status: string;
  /** The score's inputs when it exists; otherwise why it does not. */
  confidence_breakdown: Record<string, unknown>;
  selection_reason: string;
  valid_from: string;
  calculated_at: string;
  algorithm_version: string;
  supporting_count: number;
  dissenting_count: number;
  source_count: number;
}

/** One observation's role in a state. The provenance trail. */
export interface Evidence {
  observation_id: string;
  source_id: string;
  stream_id: string;
  external_id: string;
  role: "supporting" | "dissenting" | "excluded" | "considered";
  weight: string;
  observed_value: unknown;
  /** When the source says it was true. */
  event_time: string;
  /** When RealitySync learned it. Deliberately separate from `event_time`. */
  ingested_at: string;
  exclusion_reason: string | null;
}

export interface RecalculateResult {
  entity_id: string;
  attributes_considered: number;
  states_written: number;
  conflicts_written: number;
  calculated_at: string;
  /** How many written states carry no confidence score. */
  states_unscored: number;
  unscored_attributes: { attribute: string; blocked_on: string }[];
  /**
   * True when nothing could be scored. It does NOT mean nothing was written —
   * states are written either way now, with a null confidence.
   */
  blocked: boolean;
  blocked_on: string[];
  missing_specifications: { name: string; description: string }[];
}

export function listEntities(): Promise<Entity[]> {
  return apiFetch<Entity[]>("/api/entities", { cache: "no-store" });
}

export function createEntity(input: {
  entity_type: string;
  natural_key: string;
  display_name?: string;
}): Promise<Entity> {
  return apiFetch<Entity>("/api/entities", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function listConflicts(
  params: { status?: ConflictStatus } = {},
): Promise<Conflict[]> {
  const query = params.status ? `?status=${params.status}` : "";
  return apiFetch<Conflict[]>(`/api/conflicts${query}`, { cache: "no-store" });
}

export function updateConflict(
  id: string,
  input: { status: "acknowledged" | "resolved" | "dismissed"; note?: string },
): Promise<Conflict> {
  return apiFetch<Conflict>(`/api/conflicts/${id}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export function fetchTimeline(
  entityId: string,
  params: {
    axis?: TimeAxis;
    as_of_event_time?: string;
    as_of_knowledge_time?: string;
  } = {},
): Promise<Timeline> {
  const query = new URLSearchParams();
  if (params.axis) query.set("axis", params.axis);
  if (params.as_of_event_time)
    query.set("as_of_event_time", params.as_of_event_time);
  if (params.as_of_knowledge_time)
    query.set("as_of_knowledge_time", params.as_of_knowledge_time);

  const suffix = query.toString() ? `?${query}` : "";
  return apiFetch<Timeline>(`/api/entities/${entityId}/timeline${suffix}`, {
    cache: "no-store",
  });
}

export function listRealityStates(entityId: string): Promise<RealityState[]> {
  return apiFetch<RealityState[]>(`/api/entities/${entityId}/reality`, {
    cache: "no-store",
  });
}

export function listEvidence(
  entityId: string,
  attribute: string,
): Promise<Evidence[]> {
  return apiFetch<Evidence[]>(
    `/api/entities/${entityId}/reality/${encodeURIComponent(attribute)}/evidence`,
    { method: "GET" },
  );
}

export function recalculate(entityId: string): Promise<RecalculateResult> {
  return apiFetch<RecalculateResult>(`/api/entities/${entityId}/recalculate`, {
    method: "POST",
    timeoutMs: 60_000,
  });
}

File: apps/api/app/schemas/reality.py
"""Entity, reality, conflict and timeline API models."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, computed_field, field_validator

from app.models.conflict import ConflictStatus
from app.models.reality_state import RealityStatus

Identifier = Annotated[str, Field(min_length=1, max_length=256)]


# --- Entities --------------------------------------------------------------


class CreateEntityRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entity_type: Annotated[str, Field(min_length=1, max_length=64)]
    natural_key: Identifier
    display_name: Annotated[str, Field(max_length=256)] | None = None

    @field_validator("entity_type", "natural_key")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("Value must not be blank")
        return stripped


class CreateMappingRequest(BaseModel):
    """Declare that a source row describes this entity.

    Deliberately explicit: MVP entity resolution is a human decision, and this
    request is that decision being recorded.
    """

    model_config = ConfigDict(extra="forbid")

    stream_id: uuid.UUID
    external_id: Annotated[str, Field(min_length=1, max_length=512)]


class EntityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entity_type: str
    natural_key: str
    display_name: str | None
    mapping_count: int = 0
    observation_count: int = 0
    created_at: datetime


class MappingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entity_id: uuid.UUID
    stream_id: uuid.UUID
    external_id: str
    created_at: datetime


# --- Reality state ---------------------------------------------------------


class EvidenceResponse(BaseModel):
    """One observation's contribution. The provenance trail.

    Carries enough to answer "why does the system believe that" without a
    second request: which observation, from which source, what it said, when it
    was true, when we learned it, and what part it played. Anything less makes
    the caller join it back together themselves, and a provenance trail that
    needs assembling is one nobody checks.
    """

    observation_id: uuid.UUID
    #: Which source said it. The identity, not the credential.
    source_id: uuid.UUID
    stream_id: uuid.UUID
    external_id: str
    role: str
    #: Zero for every unscored state, meaning "not weighted" rather than
    #: "weighed and found worthless". The state's null confidence carries that
    #: distinction; this field cannot.
    weight: Decimal
    observed_value: Any = None
    #: When the source says it was true.
    event_time: datetime
    #: When RealitySync learned it. Separate on purpose — the two answer
    #: different questions and conflating them erases late arrival.
    ingested_at: datetime
    exclusion_reason: str | None = None


class RealityStateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entity_id: uuid.UUID
    attribute: str
    #: None when no value was selected. ``value_selected`` distinguishes that
    #: from a source genuinely asserting JSON null.
    value: Any
    value_selected: bool
    #: None while the scoring specification is unavailable. Never 0 — a client
    #: must render "unavailable", not "0%".
    confidence: Decimal | None
    status: RealityStatus
    #: Every input to the score when one exists; otherwise the reason there is
    #: none, including which specifications are outstanding.
    confidence_breakdown: dict[str, Any]
    selection_reason: str
    valid_from: datetime
    valid_to: datetime | None
    calculated_at: datetime
    algorithm_version: str
    supporting_count: int
    dissenting_count: int
    source_count: int

    @computed_field  # type: ignore[prop-decorator]
    @property
    def confidence_available(self) -> bool:
        """Explicit flag so a client branches on intent, not on a null.

        A missing field and a null field are easy to conflate with a
        serialisation bug; a boolean that says "no score exists" is not.
        """
        return self.confidence is not None


class UnscoredAttributeResponse(BaseModel):
    """What is known about an attribute when scoring is unavailable.

    Not a reality state and deliberately not shaped like one: no value, no
    confidence, no status. It reports which distinct values the sources assert
    and whether they disagree — facts that need no formula — and nothing more.
    """

    attribute: str
    scored: Literal[False] = False
    disagreement: bool
    divergence: str | None = None
    distinct_values: list[dict[str, Any]] = Field(default_factory=list)
    excluded: list[dict[str, Any]] = Field(default_factory=list)


class RecalculateResponse(BaseModel):
    """Outcome of a recalculation run."""

    entity_id: uuid.UUID
    attributes_considered: int
    states_written: int
    conflicts_written: int
    calculated_at: datetime
    #: How many written states carry no confidence score. States are written
    #: either way now — Phase 5 wrote none when scoring was blocked, which left
    #: the Reality page indistinguishable from "no data".
    states_unscored: int = 0
    #: Which attributes those are, and what each is blocked on.
    unscored_attributes: list[dict[str, str]] = Field(default_factory=list)
    #: True when no state could be scored. Kept from Phase 5 so existing
    #: clients continue to work; it no longer means "nothing was written".
    blocked: bool = False
    blocked_on: list[str] = Field(default_factory=list)
    missing_specifications: list[dict[str, str]] = Field(default_factory=list)


# --- Conflicts -------------------------------------------------------------


class ConflictResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entity_id: uuid.UUID
    entity_natural_key: str | None = None
    reality_state_id: uuid.UUID | None
    attribute: str
    conflict_type: str
    #: "unspecified" while the severity thresholds are missing. Not graded as
    #: low, which would read as "harmless".
    severity: str
    status: ConflictStatus
    #: NULL while the conflict-score formula is missing.
    score: Decimal | None
    summary: str
    details: dict[str, Any]
    detected_at: datetime
    last_seen_at: datetime
    resolved_at: datetime | None
    resolution_note: str | None


class UpdateConflictRequest(BaseModel):
    """Move a conflict through its lifecycle.

    Resolution is a human act. The engine never sets these — it reports what it
    sees, and a person decides what to do about it.
    """

    model_config = ConfigDict(extra="forbid")

    status: Literal["acknowledged", "resolved", "dismissed"]
    note: Annotated[str, Field(max_length=2000)] | None = None


# --- Timeline --------------------------------------------------------------


class TimelineEventResponse(BaseModel):
    observation_id: uuid.UUID
    external_id: str
    source_id: uuid.UUID
    source_name: str
    values: dict[str, Any]
    #: When the fact was true, per the source.
    event_time: datetime
    #: When RealitySync learned it.
    ingested_at: datetime
    event_time_semantics: str
    #: True when learned after it was true — the signal that the two axes have
    #: diverged for this record.
    arrived_late: bool
    lag_seconds: float


class TimelineResponse(BaseModel):
    """A bitemporal reconstruction, with the parameters that produced it.

    The parameters are returned because a timeline read without knowing which
    axis produced it is uninterpretable — two views of the same entity can
    legitimately differ.
    """

    axis: Literal["event", "knowledge"]
    as_of_event_time: datetime | None
    as_of_knowledge_time: datetime | None
    event_count: int
    late_arrival_count: int
    truncated: bool
    events: list[TimelineEventResponse]


class HistoricalAttributeResponse(BaseModel):
    """One field as it stood at a chosen moment."""

    model_config = ConfigDict(from_attributes=True)

    attribute: str
    status: str
    value: Any
    value_selected: bool
    confidence: Decimal | None
    confidence_available: bool
    selection_reason: str
    supporting_count: int
    dissenting_count: int
    source_count: int
    candidate_count: int


class RealityAsOfResponse(BaseModel):
    """What RealitySync would have said at ``known_at``.

    ``observations_since`` is the interesting number: it counts records that
    exist now but had not arrived by then. When a past answer differs from
    today's and no source changed its mind, this is why.
    """

    model_config = ConfigDict(from_attributes=True)

    entity_id: uuid.UUID
    known_at: datetime
    observations_known: int
    observations_since: int
    attributes: list[HistoricalAttributeResponse]
